vispy/color/colormap.py

```diff
@@ -116,44 +119,78 @@ def step(colors, x, controls=None):
+
+    controls : list
+        The list of control points for the given colors. It should be
+        an increasing list of floating-point number between 0.0 and 1.0.
+        The first control point must be 0.0. The last control point must be
+        1.0. The number of control points depends on the interpolation scheme.
+
+    texture_map_data : ndarray, shape(LUT_len, 4)
+        Numpy array of size of 1D texture lookup data
+        for luminance to RGBA conversion.
+
+    """
     assert (controls[0], controls[-1]) == (0., 1.)
     ncolors = len(controls)
     assert ncolors >= 2
-    if ncolors == 2:
-        s = "    return mix($color_0, $color_1, t);\n"
-    else:
-        s = ""
-        for i in range(ncolors-1):
-            if i == 0:
-                ifs = 'if (t < %.6f)' % (controls[i+1])
-            elif i == (ncolors-2):
-                ifs = 'else'
-            else:
-                ifs = 'else if (t < %.6f)' % (controls[i+1])
-            adj_t = '(t - %s) / %s' % (controls[i],
-                                       controls[i+1] - controls[i])
-            s += ("%s {\n    return mix($color_%d, $color_%d, %s);\n} " %
-                  (ifs, i, i+1, adj_t))
-    return "vec4 colormap(float t) {\n%s\n}" % s
-
-
-def _glsl_step(controls=None):
+    assert (texture_map_data is not None)
+
+    LUT = texture_map_data
+    LUT_len = texture_map_data.shape[0]
+
+    # Perform linear interpolation for each RGBA color component.
+    c_rgba = ColorArray(colors)._rgba
+    x = np.linspace(0.0, 1.0, LUT_len)
+    LUT[:, 0, 0] = np.interp(x, controls, c_rgba[:, 0])
+    LUT[:, 0, 1] = np.interp(x, controls, c_rgba[:, 1])
+    LUT[:, 0, 2] = np.interp(x, controls, c_rgba[:, 2])
+    LUT[:, 0, 3] = np.interp(x, controls, c_rgba[:, 3])
+
+    s2 = "uniform sampler2D texture2D_LUT;"
+    s = "{\n return texture2D(texture2D_LUT, \
+          vec2(0.0, clamp(t, 0.0, 1.0)));\n} "
+
+    return "%s\nvec4 colormap(float t) {\n%s\n}" % (s2, s)
+
+
+def _glsl_step(controls=None, colors=None, texture_map_data=None):
     assert (controls[0], controls[-1]) == (0., 1.)
     ncolors = len(controls) - 1
     assert ncolors >= 2
-    s = ""
-    for i in range(ncolors-1):
-        if i == 0:
-            ifs = 'if (t < %.6f)' % (controls[i+1])
-        elif i == (ncolors-2):
-            ifs = 'else'
-        else:
-            ifs = 'else if (t < %.6f)' % (controls[i+1])
-        s += """%s {\n    return $color_%d;\n} """ % (ifs, i)
-    return """vec4 colormap(float t) {\n%s\n}""" % s
+    assert (texture_map_data is not None)
+
+    LUT = texture_map_data
+    LUT_len = texture_map_data.shape[0]
+    LUT_tex_idx = np.linspace(0.0, 1.0, LUT_len)
+
+# Replicate indices to colormap texture and color control points.
+# The resulting matrices have size of (LUT_len,len(controls)).
+    t2 = np.repeat(LUT_tex_idx[:, np.newaxis], len(controls), 1)
+    controls2 = np.repeat(controls[:, np.newaxis], LUT_len, 1).transpose()
```

Comment:
Assuming I understand this code, it would be better to avoid `np.repeat` and use broadcasting to implicitly do it for you with something like:
```
bn = np.sum(LUT_tex_idx[:, np.newaxis] >= t2[np.newaxis], axis=1)
```